File: go/leetcode/345. Reverse Vowels of a String.go
func reverseVowels(s string) string {
    str := []uint8(s)
    vowels := map[byte]byte{
        		byte('a'): byte('a'),
				byte('e'): byte('e'),
				byte('i'): byte('i'),
				byte('o'): byte('o'),
				byte('u'): byte('u'),
				byte('A'): byte('A'),
				byte('E'): byte('E'),
				byte('I'): byte('I'),
				byte('O'): byte('O'),
				byte('U'): byte('U'),

    }
    start := 0
    end := len(s) - 1
    
    for start < end {
        _, ok := vowels[s[start]] 
        _, fine := vowels[s[end]]
        
        if ok && fine {
            temp := str[start]
            str[start] = str[end]
            str[end] = temp
            start++
            end--
            continue
        } 
        if ok != true {
            start++
        }
        if fine != true {
            end--
        }
    }
    
    return string(str)
}

File: python/leetcode/345. Reverse Vowels of a String.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        vowels = ['a', 'e', 'i', 'o', 'u']
        start = 0
        end = len(s) - 1

        while start < end:
            if s[start].casefold() in vowels and s[end].casefold() in vowels:
                st = s[start]
                en = s[end]
                s = s[:start] + en + s[start + 1:end] + st + s[end + 1:]
                start += 1
                end -= 1
                continue

            if s[end].casefold() not in vowels:
                end -= 1
            if s[start].casefold() not in vowels:
                start += 1

            print(start, end, len(s))

        return s
